### ISSUE TITLE
Incorrect Implementation of the Paterson Condition

### ISSUE THREAD

**Issue by ekpyron** [ethlts/solcore#23]
Reproduction:
```
class self:A {}
class self:B {}
class self:C {}
class self:D {}


data Uint256 = U
data T(x) = T
data S(x) = T

// This works.
instance (U:A) => T(U):D {}

// This should also work, but reports a violation of the Paterson condition.
instance (U:A, U:B, U:C) => S(U):D {}
```

The Paterson condition according to https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/instances.html#instance-termination-rules is stated as follows:

The Paterson Conditions: for each class constraint (C t1 ... tn) in the context
 1. No type variable has more occurrences in the constraint than in the head
 2. The constraint has fewer constructors and variables (taken together and counting repetitions) than the head
 3. The constraint mentions no type functions. A type function application can in principle expand to a type of arbitrary size, and so are rejected out of hand

Note that this is stated as **for each class constraint (C t1 ... tn)** in the context - not for the context in its entirety.
However, from the looks of it, our implementation sums up constructors and variables over *all* constraints in the context.

**Comment by mbenke** [comment]
Fix proposed in #22

**Comment by rodrigogribeiro** [comment]
Fixed and reviewed.
